### PR TITLE
Fix kernel timer: use after free

### PR DESCRIPTION
**Pull request by ax3l**
Fixing a bug spotted by @WeiqunZhang: the `amrex::ParallelFor` functions are async. Thus, the free after their call can lead to a use-after-free of the cost result in the still running kernels.

We think we see that at larger scale runs at the moment.

This also guards the allocation and sync to the use-cases when the GPUTimes are enabled.

### CODE REVIEW

File: Source/Particles/Deposition/CurrentDeposition.H [hunk 1]
```diff
@@ -386,12 +391,13 @@ void doEsirkepovDepositionShapeN (const GetParticlePosition& GetPosition,
     Real const clightsq = 1.0_rt / ( PhysConst::c * PhysConst::c );
 
     // Loop over particles and deposit into Jx_arr, Jy_arr and Jz_arr
+#if defined(WARPX_USE_GPUCLOCK)
     amrex::Real* cost_real = (amrex::Real*) amrex::The_Managed_Arena()->alloc(sizeof(amrex::Real));
```
Comment (ax3l) [inline Source/Particles/Deposition/CurrentDeposition.H:395]:
```suggestion
    amrex::Real* cost_real = (amrex::Real*) amrex::The_Managed_Arena()->alloc(sizeof(amrex::Real));
    *cost_real = 0.;
```

File: Source/Particles/Deposition/CurrentDeposition.H [hunk 2]
```diff
@@ -386,12 +391,13 @@ void doEsirkepovDepositionShapeN (const GetParticlePosition& GetPosition,
...
-    *cost_real = 0.;
+#endif()
```
Comment (ax3l) [inline Source/Particles/Deposition/CurrentDeposition.H]:
```suggestion
    *cost_real = 0.;
#endif
```

File: Source/Particles/Deposition/CurrentDeposition.H [hunk 3]
```diff
@@ -964,7 +989,12 @@ void doVayDepositionShapeN (const GetParticlePosition& GetPosition,
         }
 #endif
     } );
-    amrex::The_Managed_Arena()->free(cost_real);
+#   if defined(WARPX_USE_GPUCLOCK)
+    if( load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::GpuClock) {
+        amrex::Gpu::synchronize();
```
Comment (ax3l) [inline Source/Particles/Deposition/CurrentDeposition.H]: Not sure if stream synchronize would be sufficient (all case above)?
```suggestion
        amrex::Gpu::streamSynchronize();
```

Comment (ax3l) [reply Source/Particles/Deposition/CurrentDeposition.H:994]: I think we are lacking an add to the WarpX `cost` array here as well, between the synchronize and the free

**Review by mrowan137 — APPROVED** [review]

**Review by atmyers — APPROVED** [review]

**Pull request merged**